tpu-client-next/src/connection_workers_scheduler.rs

```diff
@@ -115,39 +128,41 @@ impl ConnectionWorkersScheduler {
                 }
             };
             let updated_leaders = leader_updater.next_leaders(lookahead_slots);
-            let new_leader = &updated_leaders[0];
-            let future_leaders = &updated_leaders[1..];
-            if !workers.contains(new_leader) {
-                debug!("No existing workers for {new_leader:?}, starting a new one.");
-                let worker = Self::spawn_worker(
-                    &endpoint,
-                    new_leader,
-                    worker_channel_size,
-                    skip_check_transaction_age,
-                    max_reconnect_attempts,
-                );
-                workers.push(*new_leader, worker).await;
-            }
 
-            tokio::select! {
-                send_res = workers.send_transactions_to_address(new_leader, transaction_batch) => match send_res {
-                    Ok(()) => (),
-                    Err(WorkersCacheError::ShutdownError) => {
-                        debug!("Connection to {new_leader} was closed, worker cache shutdown");
-                    }
-                    Err(err) => {
-                        warn!("Connection to {new_leader} was closed, worker error: {err}");
-                       // If we has failed to send batch, it will be dropped.
-                    }
-                },
-                () = cancel.cancelled() => {
-                    debug!("Cancelled: Shutting down");
-                    break;
+            let (new_leaders, future_leaders) = split_leaders(&updated_leaders, &leaders_fanout);
+            for new_leader in new_leaders {
+                if !workers.contains(new_leader) {
+                    debug!("No existing workers for {new_leader:?}, starting a new one.");
+                    let worker = Self::spawn_worker(
+                        &endpoint,
+                        new_leader,
+                        worker_channel_size,
+                        skip_check_transaction_age,
+                        max_reconnect_attempts,
+                    );
+                    workers.push(*new_leader, worker).await;
                 }
-            };
 
-            // Regardless of who is leader, add future leaders to the cache to
-            // hide the latency of opening the connection.
+                tokio::select! {
+                    send_res = workers.send_transactions_to_address(new_leader, transaction_batch.clone()) => match send_res {
```

Comment:
Right now this will block if the corresponding worker's channel is full, so
effectively the slowest leader will slow down all other leaders.

I think we should 1) increase the channel size and 2) add a
try_send_transactions_to_address that returns an error if the channel is full.
If the channel is full, I guess as a start we could log a warning and drop the
batch for that leader, but at least we don't slow down the other leaders.

Then longer term, we need to think of a better API so that the caller of the
crate can decide what should happen: drop the batch? increase the channel size?
slow down upstream?

Reply:
This makes sense, I'm not sure that there is a value in having both `send_transactions` and `try_send_transactions` in the context of this crate. So probably just modify the code to use `try_`

Reply:
The size of the channel `worker_channel_size` is configurable through the Scheduler config

Reply:
one thing that worries me is the scenario with `transaction-bench`. There we actually need to have a backpressure (which has been created with these `worker_channel.send().await`) because it tries to generate as many as possible transaction batches and we rely on slowing down sending them.

Reply:
In case of STS, `try_send` is fine because even if we drop some transactions silently, there is a retry thread which sends them again anyways.